### PR TITLE
Improve CLI interface

### PR DESCRIPTION
**Pull request by jvegasbsc**
A first draft of an improved command line interface using subcommands. I was going to open an issue to discuss thit, but finally decided to create a small pull request so you can see what I wanted to add.

The idea is to expand the capabilities of the esmvaltool command by using subcomands, like git and lots of other programs do. The objective will be to have all esmvaltool utilities, like the cmorizer, as subcommands to make the tool more user friendly.

In this pull request I add a run command (the tool as we use it know), a list command that shows all the installed recipes (not the examples, by the way: it is not recursice) and a get that copies one of the installed recipes so users can modify them

What do you think?

**Tasks**

-   [ ] [Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do, if you haven't done so already (and add the link at the bottom)
-   [ ] This pull request has a descriptive title that can be used in a changelog
-   [ ] Add unit tests
-   [ ] Public functions should have a numpy-style docstring so they appear properly in the [API documentation](https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/api/esmvalcore.html). For all other functions a one line docstring is sufficient.
-   [ ] If writing a new/modified preprocessor function, please update the [documentation](https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/esmvalcore/preprocessor.html)
-   [ ] Circle/CI tests pass. Status can be seen below your pull request. If the tests are failing, click the link to find out why.
-   [ ] Codacy code quality checks pass. Status can be seen below your pull request. If there is an error, click the link to find out why. If you suspect Codacy may be wrong, please ask by commenting.
-   [ ] Please use `yamllint` to check that your YAML files do not contain mistakes
-   [ ] If you make backward incompatible changes to the recipe format, make a new pull request in the [ESMValTool repository](https://github.com/ESMValGroup/ESMValTool) and add the link below

If you need help with any of the tasks above, please do not hesitate to ask by commenting in the issue or pull request.

* * *

Closes {Link to corresponding issue}

### CODE REVIEW

File: esmvalcore/_main.py [hunk 1]
```diff
@@ -64,46 +65,51 @@ def get_args():
...
         help='Config file')
-    parser.add_argument(
+    run.add_argument(
         '-s',
         '--synda-download',
         action='store_true',
         help='Download input data using synda. This requires a working '
         'synda installation.')
-    parser.add_argument(
+    run.add_argument(
```
Comment (bouweandela) [inline esmvalcore/_main.py:91]: Many of the `run` settings are meant for debugging. Would it be an idea to place those in a `debug` command, to indicate clearly that those are mostly meant for diagnostic developers?

**Comment by bouweandela** [comment]
Also, do you have ideas on how we could add tests for our command line interface?

**Comment by bouweandela** [comment]
Another idea that I had was to add a `esmvaltool show examples/recipe_python.yml` command, that prints the documentation/description section from the recipe to stdout.

**Comment by bouweandela** [comment]
> not the examples, by the way: it is not recursive

It might be nicer to make it recursive, because at some point we may have so many recipes that we will have to put them in subdirectories. We already do that for the recipes created in the eWaterCycle project (they are in the recipes/hydrology folder). We could follow the same structure as on readthedocs: https://esmvaltool.readthedocs.io/en/latest/recipes/index.html

**Comment by valeriupredoi** [comment]
`Use fire` - best commit message ever :rofl:

**Comment by jvegasbsc** [comment]
> Use fire - best commit message ever 🤣

I just discovered fire, so it is compeltely normal 😆

**Comment by jvegasbsc** [comment]
Regarding that, swiching to fire makes everything a lot easier with the CLI, but we lose the ability to have single letter options. 

To avoid this becoming an usability problem, the config-user is now readed from ~/.esmvaltool/config-user.yml if no one is passed

**Comment by jvegasbsc** [comment]
> It might be nicer to make it recursive, because at some point we may have so many recipes that we will have to put them in subdirectories. We already do that for the recipes created in the eWaterCycle project (they are in the recipes/hydrology folder). We could follow the same structure as on readthedocs: https://esmvaltool.readthedocs.io/en/latest/recipes/index.html

Yes, I know. It was just a quick mock up to show how a CLI based on commands would work

File: esmvalcore/_main.py (hunk 1 above)
Comment (jvegasbsc) [reply esmvalcore/_main.py]: Maybe

**Comment by jvegasbsc** [comment]
I added a couple more functionalities to better represent what we may achieve in the future.

There is one thing I would like to do but I do not have a clear idea on how: allow ESMValTool package to add its own commands, like the cmorize_obs or the check_ncl we are providing *whitour* ESMValCore knowing anything about them. Any ideas?

**Comment by bouweandela** [comment]
I think you could just import the relevant functions/classes from the esmvaltool package and add them as attributes to the `ESMValTool` class here:
https://github.com/ESMValGroup/ESMValCore/blob/3a2fcec3b6d474969f42c213f0d61de8a5232946/esmvalcore/_main.py#L211-L213

e.g.
```python
try:
    from esmvaltool import commands
except ImportError:
    pass
else:
    for cmd in commands.__all__:
        setattr(self, cmd, getattr(commands, cmd))
```
and the file `esmvaltool/commands.py` would look like:
```python
from .cmorizers.obs.cmorize_obs import cmorize

__all__ = [
    'cmorize',
]
```

**Comment by jvegasbsc** [comment]
I found another way, that I described in ESMValGroup/ESMValTool#1657

It is using entry points, so it does not need to assume anything about the package

**Comment by Peter9192** [comment]
+1 for the `get` command. 

I wonder if this PR makes it possible/easier to change the way ESMValTool is configured. I often change the configuration settings in config-user.yml, for example the output directory or the max_parralel_tasks. It would be super nice if I could specify those as command line options instead.

**Comment by jvegasbsc** [comment]
> I wonder if this PR makes it possible/easier to change the way ESMValTool is configured. I often change the configuration settings in config-user.yml, for example the output directory or the max_parralel_tasks. It would be super nice if I could specify those as command line options instead.

Yes, it is possible... and done. The run command admits extra arguments that will be added / overwrite what you have in the config user. So you can write something like this

```bash
esmvaltool run recipe_albedolandcover.yml  --write_plots=False
```

**Comment by jvegasbsc** [comment]
Ready to test. I will update the documentation when we all agree that this is the way to go

**Comment by jvegasbsc** [comment]
> One thing I noticed is that loading the text feels quite slow, e.g. the command esmvaltool --help takes about 3 seconds before it shows any text. Do you think this could be improved?

I am going to change the top level imports to local ones in `_main.py`. This should help a bit

**Comment by jvegasbsc** [comment]
I am also playing a bit with the autocompletion features of Fire. If you want to try you need to:

 - Generate the autocompletion script by running `esmvaltool -- --completion > ~/.esmvaltool-completion`

- Source  `~/.esmvaltool-completion` from your .bashrc

**Comment by jvegasbsc** [comment]
> I am going to change the top level imports to local ones in _main.py. This should help a bit

Done. Startup time has improved a lot. Can you test, please?

**Comment by valeriupredoi** [comment]
```bash
(esmvaltool) valeriu@valeriu:~/ESMValCore$ esmvaltool run
ERROR: The function received no value for the required argument: recipe
Usage: esmvaltool run RECIPE <flags>
  optional flags:        --config_file | --max_datasets | --max_years |
                         --skip_nonexistent | --synda_download |
                         --diagnostics | --check_level
  additional flags are accepted

For detailed information on this command, run:
  esmvaltool run -- --help
```
not 100% straightforward since it needs **both** recipe and user config file :beer:

**Comment by valeriupredoi** [comment]
BTW really nice CLI @jvegasbsc - me likes it! :beer: 

This is bothersome though:
```
FLAGS
    --config_file=CONFIG_FILE
        Config file to use. If not provided will load ${HOME}/.esmvaltool/config.user.yml if it exists
```
and shortops like `-c` don't work anymore?

**Comment by jvegasbsc** [comment]
> not 100% straightforward since it needs both recipe and user config file 🍺

Not really, if no config-user is provided it will look for one  ${HOME}/.esmvaltool

**Comment by jvegasbsc** [comment]
> and shortops like -c don't work anymore?

Fire automatically set them up if there is no possible conflict with other options, in this case it conflicts with `--check_level`, so it is not available

**Comment by jvegasbsc** [comment]
>esmvaltool recipes get hydrology/recipe_wflow.yml the recursive thing you discussed above?

It is working for me, can you give me the console output?

> esmvaltool version now prints: ESMValCore: 2.0.0b9 why is this Core and not Tool?

Because esmvaltool is the main executable, and the new version can only detect the installation if you have installed  ESMValGroup/ESMValTool#1666. It will print the versions of core and every other package that has registered an entry point in `esmvaltool_commands`, which for now is only the toolif you are using this branch

> esmvaltool run hydrology/recipe_wflow.yml looked in the ESMValCore directory, fixed by also re-installing esmvaltool itself; wouldn't it make more sense for core to expose the CLI command ESMValCore?

I agree this is confusing, but I think it would be worse if the main executable of ESMValTool is called `esmvalcore`

> Furthermore, I thought the help screen looked a bit alien. It starts less, I suppose, whereas the old behaviour was just to print to the terminal? Maybe it sets me off because it's mostly an empty screen. 

I have changed it to the old behaviour. By default, Fire is using `more`

> Also I found the instructions listed here not very clear as of now, but I guess you're planning to update that later, right?

Yes

**Comment by Peter9192** [comment]
> It is working for me, can you give me the console output?

Here it is:
```shell
$ esmvaltool run --config_file=esmvaltool_config.yml examples/recipe_python.yml
...
2020-06-23 09:11:38,946 UTC [18892] ERROR   Program terminated abnormally, see stack trace below for more information
Traceback (most recent call last):
  File "~/ESMValCore/esmvalcore/_main.py", line 363, in run
    fire.Fire(ESMValTool())
  File "/home/peter9192/miniconda3/envs/evert/lib/python3.8/site-packages/fire/core.py", line 138, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
  File "/home/peter9192/miniconda3/envs/evert/lib/python3.8/site-packages/fire/core.py", line 463, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
  File "/home/peter9192/miniconda3/envs/evert/lib/python3.8/site-packages/fire/core.py", line 672, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
  File "~/ESMValCore/esmvalcore/_main.py", line 340, in run
    process_recipe(recipe_file=recipe, config_user=cfg)
  File "~/ESMValCore/esmvalcore/_main.py", line 101, in process_recipe
    recipe = read_recipe_file(recipe_file, config_user)
  File "~/ESMValCore/esmvalcore/_recipe.py", line 71, in read_recipe_file
    return Recipe(raw_recipe,
  File "~/ESMValCore/esmvalcore/_recipe.py", line 944, in __init__
    self.diagnostics = self._initialize_diagnostics(
  File "~/ESMValCore/esmvalcore/_recipe.py", line 983, in _initialize_diagnostics
    self._initialize_preprocessor_output(
  File "~/ESMValCore/esmvalcore/_recipe.py", line 1097, in _initialize_preprocessor_output
    self._initialize_variables(raw_variable, raw_datasets)
  File "~/ESMValCore/esmvalcore/_recipe.py", line 1051, in _initialize_variables
    variable['start_year'] + self._cfg['max_years'] - 1)
TypeError: unsupported operand type(s) for +: 'int' and 'NoneType'
2020-06-23 09:11:38,948 UTC [18892] INFO    If you suspect this is a bug or need help, please open an issue on https://github.com/ESMValGroup/ESMValTool/issues and attach the run/recipe_*.yml and run/main_log_debug.txt files from the output directory.
```

**Comment by jvegasbsc** [comment]
> Here it is:

Should be fixed now

**Comment by Peter9192** [comment]
> Should be fixed now

Yep, confirmed!

**Comment by jvegasbsc** [comment]
Please, test this as much as possible. The functionality should be all there, I just want to improve the documentation now

**Comment by bouweandela** [comment]
@jvegasbsc To make the review a bit faster for today, I've directly made some changes, hope that's ok with you!

File: tests/integration/test_main.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
+# """
```
Comment (bouweandela) [inline tests/integration/test_main.py:1]: Why are all the tests commented out?

Comment (jvegasbsc) [reply tests/integration/test_main.py:1]: I did not manage to make them work on time for the release.

**Pull request merged**